File: smoke_tests/ruby/ssrf.rb
# Server-Side Request Forgery (SSRF) vulnerabilities in Ruby
require 'net/http'
require 'open-uri'
require 'faraday'
require 'httparty'
require 'rest-client'

class SsrfController < ApplicationController
  # Test 1: Net::HTTP with user URL
  def fetch_url
    url = params[:url]
    uri = URI.parse(url)
    # VULNERABLE: User-controlled URL
    response = Net::HTTP.get(uri)
    render plain: response
  end

  # Test 2: open-uri with user URL
  def open_url
    url = params[:url]
    # VULNERABLE: open-uri follows redirects
    content = URI.open(url).read
    render plain: content
  end

  # Test 3: Faraday with user URL
  def faraday_fetch
    url = params[:url]
    # VULNERABLE: User-controlled URL
    response = Faraday.get(url)
    render plain: response.body
  end

  # Test 4: HTTParty with user URL
  def httparty_fetch
    url = params[:url]
    # VULNERABLE: User-controlled URL
    response = HTTParty.get(url)
    render plain: response.body
  end

  # Test 5: RestClient with user URL
  def rest_fetch
    url = params[:url]
    # VULNERABLE: User-controlled URL
    response = RestClient.get(url)
    render plain: response.body
  end

  # Test 6: Partial URL construction
  def fetch_from_host
    host = params[:host]
    # VULNERABLE: User controls hostname
    url = "http://#{host}/api/data"
    response = Net::HTTP.get(URI.parse(url))
    render plain: response
  end

  # Test 7: Port scanning via SSRF
  def check_port
    port = params[:port]
    # VULNERABLE: User controls port
    uri = URI.parse("http://internal-server:#{port}/")
    begin
      Net::HTTP.get(uri)
      render json: { status: 'open' }
    rescue
      render json: { status: 'closed' }
    end
  end

  # Test 8: Image proxy
  def image_proxy
    image_url = params[:src]
    # VULNERABLE: Can fetch internal resources
    uri = URI.parse(image_url)
    response = Net::HTTP.get_response(uri)
    send_data response.body, type: 'image/png'
  end

  # Test 9: Webhook URL
  def send_webhook
    webhook_url = params[:webhook]
    data = params[:data]
    # VULNERABLE: User-controlled webhook
    uri = URI.parse(webhook_url)
    Net::HTTP.post(uri, data.to_json)
    head :ok
  end

  # Test 10: File protocol SSRF
  def read_resource
    uri = params[:uri]
    # VULNERABLE: Could be file:///etc/passwd with open-uri
    content = URI.open(uri).read
    render plain: content
  end

  # Test 11: DNS rebinding
  def fetch_external
    domain = params[:domain]
    # VULNERABLE: DNS can resolve to internal IP
    url = "http://#{domain}/data"
    response = Net::HTTP.get(URI.parse(url))
    render plain: response
  end
end

File: smoke_tests/csharp/ssrf.cs
// Server-Side Request Forgery (SSRF) vulnerabilities in C#
using System;
using System.Net;
using System.Net.Http;
using System.IO;
using System.Threading.Tasks;
using System.Web.Mvc;

namespace VulnerableApp
{
    public class SsrfController : Controller
    {
        private readonly HttpClient _httpClient = new HttpClient();

        // Test 1: Direct URL from user
        public async Task<ActionResult> Fetch()
        {
            string url = Request.QueryString["url"];
            // VULNERABLE: User-controlled URL
            var response = await _httpClient.GetStringAsync(url);
            return Content(response);
        }

        // Test 2: WebClient with user URL
        public ActionResult Download()
        {
            string url = Request.QueryString["url"];
            using (var client = new WebClient())
            {
                // VULNERABLE: User-controlled download URL
                string content = client.DownloadString(url);
                return Content(content);
            }
        }

        // Test 3: WebRequest with user URL
        public ActionResult FetchData()
        {
            string url = Request.QueryString["target"];
            // VULNERABLE: User-controlled WebRequest
            WebRequest request = WebRequest.Create(url);
            using (var response = request.GetResponse())
            using (var reader = new StreamReader(response.GetResponseStream()))
            {
                return Content(reader.ReadToEnd());
            }
        }

        // Test 4: Partial URL construction
        public async Task<ActionResult> FetchFromHost()
        {
            string host = Request.QueryString["host"];
            // VULNERABLE: User controls hostname
            string url = $"http://{host}/api/data";
            var response = await _httpClient.GetStringAsync(url);
            return Content(response);
        }

        // Test 5: Port scanning via SSRF
        public async Task<ActionResult> CheckPort()
        {
            string port = Request.QueryString["port"];
            // VULNERABLE: User controls port, can scan internal network
            string url = $"http://internal-server:{port}/";
            try
            {
                var response = await _httpClient.GetAsync(url);
                return Json(new { status = "open" });
            }
            catch
            {
                return Json(new { status = "closed" });
            }
        }

        // Test 6: Image proxy
        public async Task<ActionResult> ImageProxy()
        {
            string imageUrl = Request.QueryString["src"];
            // VULNERABLE: Can fetch internal resources
            var bytes = await _httpClient.GetByteArrayAsync(imageUrl);
            return File(bytes, "image/png");
        }

        // Test 7: Webhook URL
        public async Task<ActionResult> SendWebhook()
        {
            string webhookUrl = Request.Form["webhook"];
            string data = Request.Form["data"];
            // VULNERABLE: User-controlled webhook destination
            var content = new StringContent(data);
            await _httpClient.PostAsync(webhookUrl, content);
            return Ok();
        }

        // Test 8: DNS rebinding vulnerable
        public async Task<ActionResult> FetchExternal()
        {
            string domain = Request.QueryString["domain"];
            // VULNERABLE: DNS can resolve to internal IP
            string url = $"http://{domain}/data";
            var response = await _httpClient.GetStringAsync(url);
            return Content(response);
        }

        // Test 9: File protocol SSRF
        public ActionResult ReadResource()
        {
            string uri = Request.QueryString["uri"];
            // VULNERABLE: Could be file:///etc/passwd
            WebRequest request = WebRequest.Create(uri);
            using (var response = request.GetResponse())
            using (var reader = new StreamReader(response.GetResponseStream()))
            {
                return Content(reader.ReadToEnd());
            }
        }

        // Test 10: Redirect following
        public async Task<ActionResult> FetchWithRedirect()
        {
            string url = Request.QueryString["url"];
            var handler = new HttpClientHandler
            {
                // VULNERABLE: Follows redirects to internal resources
                AllowAutoRedirect = true,
                MaxAutomaticRedirections = 10
            };
            using (var client = new HttpClient(handler))
            {
                var response = await client.GetStringAsync(url);
                return Content(response);
            }
        }
    }
}
